### PR TITLE
A terrible hash map implementation

### PR DESCRIPTION
**Pull request by SquidDev**
- Add several type classes for "indexable" types.
   - `index`: This provides `.()` and `.?()` getters.
   - `mut_index`: This provides a mutating setter `.[]<-`.
   - Add a `set_index` type class. This provides a setter `.()<-`, which returns an updated collection.
 - Add `hashable` type class (in `data/hash.ml`), and instances for the basic types.
 - Add a primitive mutable hash map (in `data/hash/map.ml`). There's a lot of work which could be done to improve this, but it's a good enough start for basic functionality.

### CODE REVIEW

**Comment by plt-abigail** [comment]
I wonder if a HAMT implementation wouldn't be better?

**Comment by SquidDev** [comment]
> But I guess it doesn't matter if everything's out of bounds for it..

Yeah, I'm _fairly_ sure this is a safe optimisation as empty array can't be modified or indexed. Java does it, though I realise that's hardly a great commendation.

> I wonder if a HAMT implementation wouldn't be better?

I don't know if it's worth having both? I definitely want some kind of mutable map structure, though am not exactly wedded to this particular design - might try out something closer to Lua's hash implementation.

**Comment by davidgarland** [comment]
If performance is your aim, I might recommend looking into [robin hood hashing](https://www.sebastiansylvan.com/post/robin-hood-hashing-should-be-your-default-hash-table-implementation/) with [backwards shift deletion](http://codecapsule.com/2013/11/17/robin-hood-hashing-backward-shift-deletion/), more or less keeping the current model of using a flat array. (Though, I think you ought to make the key/value pair into two individual arrays for the usual SoA benefits-- `<$>` on hashmaps would then no longer need to load the keys into cache as well, for instance, and the same goes for iterating over keys only.)

**Comment by SquidDev** [comment]
Thanks for the links David! I was vaguely aware of robin hood hashing, but those are articles were very useful.

I must confess this implementation was largely written from whatever I remember of algorithms courses, so is definitely not as sane as it should be. I definitely want to replace this with a better implementation.

That said, performance is a bit tricker - the backend spits out absolutely terrible Lua code, which means any memory optimisations are overshadowed by the number of closures/tables we're allocating for every call. I really need to fix this before doing any serious profiling:

```lua
-- map.[1] compiles to:
local tmp0 = _dot_28_290({
  hash = function(x) return x end,
  ["Hashable$agcy"] = {
    ["=="] = int_eq0,
    ["<>"] = function(bqq)
      return function(bqr)
        return bqq ~= bqr
      end
    end
  }
})(map)(1)
```

Looks like it's time to do some optimiser work again :D:.

File: lib/data/hash.ml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+open import "../prelude.ml"
+
+class eq 'a => hashable 'a
```
Comment (plt-hokusai) [inline lib/data/hash.ml:3]: `begin`/`end` please

File: lib/data/hash.ml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  val hash : 'a -> int
+
+instance hashable int
```
Comment (plt-hokusai) [inline lib/data/hash.ml:6]: ditto

File: lib/data/hash.ml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  let hash x = x
+
+instance hashable ()
```
Comment (plt-hokusai) [inline lib/data/hash.ml:9]: ditto

File: lib/data/hash.ml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  let hash () = 0
+
+instance hashable bool
```
Comment (plt-hokusai) [inline lib/data/hash.ml:12]: ditto

File: lib/data/hash.ml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  let hash x = if x then 1 else 0
+
+instance hashable 'a * hashable 'b => hashable ('a * 'b)
```
Comment (plt-hokusai) [inline lib/data/hash.ml:15]: ditto

File: lib/data/hash.ml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  let hash (a, b) = 31 * hash a + hash b
+
+instance hashable string
+  let hash xs =
+    (* Hash using the djb2 hash function *)
+    let open import "../lua/string.ml"
+    let rec go x = function
+    | 0 -> x
```
Comment (plt-hokusai) [inline lib/data/hash.ml:23]: Shouldn't we take the remainder of this by a suitable prime?

Comment (plt-hokusai) [reply lib/data/hash.ml:23]: Moreover, this algorithm works to hash any Foldable, does it not? We should at least add instances for lists.

File: lib/data/array.ml [hunk 7]
```diff
@@ -111,14 +118,43 @@ instance eq 'a => eq (array 'a) begin
...
 
-let ( .() ) (Array r as arr) i =
-  in_bounds "(.())" arr i
-  geti r.backing (i + r.offset)
-
-let ( .()<- ) (Array r as arr) i x =
-  in_bounds "update" arr i
-  let _ = seti r.backing (i + r.offset) x
-  ()
+let empty = Array { length = 0, offset = 0, backing = empty_storage }
```
Comment (plt-hokusai) [inline lib/data/array.ml:121]: I really don't feel safe sharing a single table as backing storage for a module of mutable arrays.

File: lib/data/hash/map.ml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      | Nil -> None
+      | Cons { k = k', v } when k == k' -> Some v
+      | Cons { next } -> find next
+    let h = (hash k .&. (Array.size array - 1)) + 1
+    find (array.[h])
+
+let iter f (HashMap container) =
+  let Container { array } = !container
+  let rec iter_one = function
+  | Nil -> ()
```
Comment (plt-hokusai) [inline lib/data/hash/map.ml:47]: Could you indent these arms?

File: lib/data/hash/map.ml [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+let load_factor = 2
+
+let make () = HashMap (ref (Container { array = Array.empty, size = 0, initial = 16 }))
+
+let make_with_capacity c =
+  let initial = if c <= 1 then 1 else
+    Math.log (float_of_int c) /. Math.log 2.0 |> Math.ceil |> (2**)
+  HashMap (ref (Container { array = Array.empty, size = 0, initial }))
+
+instance hashable 'k => index (t 'k 'v)
```
Comment (plt-hokusai) [inline lib/data/hash/map.ml:30]: Begin/end here too!

File: lib/data/hash/map.ml [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  | Nil
+  | Cons of { k : 'k, v : 'v, next : pairs 'k 'v }
+
+private type container 'k 'v = Container of {
+  array : Array.array (pairs 'k 'v),
+  size : int,
+  initial : int,
+}
+
+type t 'k 'v = private HashMap of ref (container 'k 'v)
```
Comment (plt-hokusai) [inline lib/data/hash/map.ml:19]: This type is missing instances for `functor`, `foldable`, and `traversable`.

File: lib/data/map.ml [hunk 11]
```diff
@@ -87,17 +88,24 @@ let lookup k (M tree) =
       | Lt -> go l
       | Gt -> go r
   go tree
-let (.?[]) map key = lookup key map
 
-(** Lookup a key on the map. If it is not present, raise an exception.
- *)
-let (.[]) map key =
-  match lookup key map with
+(* instance ord index map where *)
```
Comment (plt-hokusai) [inline lib/data/map.ml:92]: Huh?

File: lib/prelude.ml [hunk 12]
```diff
@@ -14,7 +14,7 @@ include import "./data/enumeration.ml"
 
 module Io = import { lua = "./lua/io.ml" }
 module Bit = import { lua = "./lua/bit.ml" }
-open Bit.Ops
+include Bit.Ops
 
 open import { lua = "./lua/conversion.ml" }
```
Comment (plt-hokusai) [inline lib/prelude.ml:19]: Include this one too, if you don't mind.

File: src/Backend/Lua/Emit.hs [hunk 13]
```diff
@@ -835,3 +835,5 @@ opOfIntrinsic StrConcat = ".."
 opOfIntrinsic StrEq = "=="
 opOfIntrinsic StrLt = "<"
 opOfIntrinsic StrLe = "<="
+
+opOfIntrinsic BoolEq = "=="
```
Comment (plt-hokusai) [inline src/Backend/Lua/Emit.hs:839]: I wonder what code we would generate for
```ocaml
let a == b =
  match a, b with
  | true, true | false, false -> true
  | _ -> true
```
(no need to act on this)

File: lib/amulet/base.ml [hunk 14]
```diff
@@ -29,6 +29,7 @@ end
 instance eq int    begin let (==) = Base_defs.int_eq end
 instance eq float  begin let (==) = Base_defs.float_eq end
 instance eq string begin let (==) = Base_defs.string_eq end
+instance eq bool   begin let (==) = Base_defs.bool_eq end
```
[the remark this answers is not in the document]
Comment (plt-hokusai) [reply lib/amulet/base.ml:32]: Nvm - I saw the primitive way down there. But `ord bool` is still good to have!

File: lib/data/array.ml (hunk 7 above)
Comment (SquidDev) [reply lib/data/array.ml:121]: I get that, but it is perfectly safe - you can't mutate it as it's empty.

File: lib/data/array.ml [hunk 15]
```diff
@@ -111,14 +118,43 @@ instance eq 'a => eq (array 'a) begin
...
+
+let iter f (Array { length, offset, backing }) =
```
[the remark this answers is not in the document]
Comment (SquidDev) [reply lib/data/array.ml:123]: Hmm. Any reason why? OCaml generally has the object you act on as the last argument, as it's more amenable to currying/`|>`.

File: lib/data/hash/map.ml (hunk 10 above)
Comment (SquidDev) [reply lib/data/hash/map.ml:19]: Good point. I originally left it off, as functors on mutable types feels a little  weird, but we also do in for arrays after all.

File: lib/amulet/base.ml (hunk 14 above)
[the remark this answers is not in the document]
Comment (SquidDev) [reply lib/amulet/base.ml:32]: Yeah, the generated code will be bad (similarly for any enum type I guess). Hence we cheating with an intrinsic for now.

File: lib/data/array.ml [hunk 16]
```diff
@@ -10,11 +11,14 @@ type array 'a =
     backing : storage 'a
   }
 
-let private in_bounds str (Array r) i =
-  if (r.offset + i <= r.length) && (i >= 1) then
-    ()
-  else
-    throw (Invalid (str ^ ": index " ^ show i ^ " is out of bounds"))
+let private is_in_bounds i (Array { offset, length }) =
+  (* Flip the argument order so the pattern match happens last. Otherwise this
```
[the remark this answers is not in the document]
Comment (SquidDev) [reply lib/data/array.ml:15]: The desugaring is correct, it's just the optimiser is dumb. ` is_in_bounds (Array { offset, length }) i` desugars to:

```ml
let is_in_bounds = λ a. 
  let a' = coerce a
  match a' with
  | { offset, length } -> λ i. (* ... *)
```

And the uncurrying optimiser isn't smart enough to deal with this.

File: lib/data/array.ml (hunk 15 above)
[the remark this answers is not in the document]
Comment (SquidDev) [reply lib/data/array.ml:123]: Also, given `iter f x` is just `foldl (fun () -> f) () x`, it feels more consistent to keep it before.

File: lib/data/array.ml (hunk 16 above)
[the remark this answers is not in the document]
Comment (plt-hokusai) [reply lib/data/array.ml:15]: Right, I feel we should float all the lambdas outwards so we don't end up with this "problem", especially for functions defined in a single block like that

File: lib/data/array.ml (hunk 15 above)
[the remark this answers is not in the document]
Comment (plt-hokusai) [reply lib/data/array.ml:123]: So we can write
```
iter x @@ fun e ->
   ...
```

File: lib/data/array.ml (hunk 7 above)
Comment (plt-hokusai) [reply lib/data/array.ml:121]: Aye, if you say so!

File: lib/data/array.ml (hunk 15 above)
[the remark this answers is not in the document]
Comment (SquidDev) [reply lib/data/array.ml:123]: `x |> iter @@ fun e ->` should work correctly.

File: lib/data/hash/map.ml (hunk 10 above)
Comment (plt-hokusai) [reply lib/data/hash/map.ml:18]: foldable` at least makes a lot of sense, but I think a copying `map` is worth having, at least.

File: lib/data/array.ml (hunk 15 above)
[the remark this answers is not in the document]
Comment (plt-hokusai) [reply lib/data/array.ml:123]: Well, if the parser agrees with this now—Last I tried, I ran into a bunch of context weirdness. Admittedly it was quite a while ago.

File: lib/data/hash/map.ml [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+private module Math = import "../../lua/math.ml"
+private module Array = import "../array.ml"
+
+private type pairs 'k 'v =
+  | Nil
+  | Cons of { k : 'k, v : 'v, next : pairs 'k 'v }
+
+instance functor (pairs 'k) begin
+  let f <$> x =
+    match x with
```
Comment (plt-hokusai) [inline lib/data/hash/map.ml:14]: Sorry to nitpick like this, but do you think you could rewrite this with a local worker? To make the dictionary non-recursive

File: lib/data/index.ml [hunk 18]
```diff
@@ -0,0 +1,43 @@
+open import "../prelude.ml"
+private module String = import "../lua/string.ml"
+
+class index 'a
```
Comment (plt-hokusai) [inline lib/data/index.ml:4]: missing begin/end

File: lib/data/index.ml [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+  val ( .() ) : 'a -> key 'a -> value 'a
+  let ( .() ) xs k =
+    match xs.?(k) with
+    | Some x -> x
+    | None -> Exception.(throw (Invalid ".()"))
+
+  (** Look up an index, returning [None] if it is unavailable. *)
+  val ( .?() ) : 'a -> key 'a -> option (value 'a)
+
+instance index string
```
Comment (plt-hokusai) [inline lib/data/index.ml:18]: ditto

File: lib/data/index.ml [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    String.char_code_at str i |> Some
+
+  let ( .() ) str i =
+    if i <= 1 || i > String.length str then
+      let open Exception in
+      throw (Invalid ("string.(): index " ^ show i ^ " is out of bounds"))
+    else
+      String.char_code_at str i
+
+class index 'a => set_index 'a
```
Comment (plt-hokusai) [inline lib/data/index.ml:33]: ditto

File: lib/data/index.ml [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+  (** Create a copy of this object with the index added or modified. *)
+  val ( .()<- ) : 'a -> key 'a -> value 'a -> 'a
+
+class index 'a => mut_index 'a
```
Comment (plt-hokusai) [inline lib/data/index.ml:37]: ditto

**Pull request merged**